// cockatiel/dist/Policy.test.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const sinon_1 = require("sinon");
const Breaker_1 = require("./breaker/Breaker");
const Errors_1 = require("./errors/Errors");
const Policy_1 = require("./Policy");
const TimeoutPolicy_1 = require("./TimeoutPolicy");
class MyError1 extends Error {
}
class MyError2 extends Error {
}
class MyError3 extends Error {
}
const assertNever = (_value) => {
    throw new Error('unreachable');
};
describe('Policy', () => {
    it('wraps', async () => {
        const policy = (0, Policy_1.wrap)((0, Policy_1.retry)((0, Policy_1.handleType)(MyError1), { maxAttempts: 3 }), (0, Policy_1.circuitBreaker)(Policy_1.handleAll, { halfOpenAfter: 100, breaker: new Breaker_1.ConsecutiveBreaker(2) }));
        // should retry and break the circuit
        await (0, chai_1.expect)(policy.execute((0, sinon_1.stub)().throws(new MyError1()))).to.be.rejectedWith(Errors_1.BrokenCircuitError);
    });
    it('wraps and keeps correct types', async () => {
        const policies = [
            (0, Policy_1.retry)(Policy_1.handleAll, { maxAttempts: 2 }),
            (0, Policy_1.circuitBreaker)(Policy_1.handleAll, { halfOpenAfter: 100, breaker: new Breaker_1.ConsecutiveBreaker(2) }),
            (0, Policy_1.fallback)(Policy_1.handleAll, 'foo'),
            (0, Policy_1.timeout)(1000, TimeoutPolicy_1.TimeoutStrategy.Aggressive),
            Policy_1.noop,
        ];
        const policy = (0, Policy_1.wrap)(...policies);
        (0, chai_1.expect)(policy.wrapped).to.deep.equal(policies);
        const result = await policy.execute(context => {
            (0, chai_1.expect)(context.signal).to.be.an.instanceOf(AbortSignal);
            (0, chai_1.expect)(context.attempt).to.equal(0);
            return 1234;
        });
        switch (typeof result) {
            case 'string':
                result.toUpperCase();
                break;
            case 'number':
                Math.pow(result, 2);
                break;
            default:
                assertNever(result);
        }
    });
    it('applies error filters', async () => {
        const fn = (0, sinon_1.stub)()
            .onCall(0)
            .throws(new MyError1())
            .onCall(1)
            .throws(new MyError2())
            .onCall(2)
            .throws(new MyError3('foo'))
            .onCall(3)
            .throws(new Error('potato'))
            .onCall(4)
            .throws(new MyError3('bar'));
        await (0, chai_1.expect)((0, Policy_1.retry)((0, Policy_1.handleType)(MyError1)
            .orType(MyError2)
            .orType(MyError3, e => e.message === 'foo')
            .orWhen(e => e.message === 'potato'), { maxAttempts: 10 }).execute(fn)).to.be.rejectedWith(MyError3, 'bar');
        (0, chai_1.expect)(fn).to.have.callCount(5);
    });
    it('applies result filters', async () => {
        const fn = (0, sinon_1.stub)()
            .onCall(0)
            .returns(new MyError1())
            .onCall(1)
            .returns(new MyError2())
            .onCall(2)
            .returns(new MyError3('foo'))
            .onCall(3)
            .returns('potato')
            .onCall(4)
            .returns('ok!');
        (0, chai_1.expect)(await (0, Policy_1.retry)((0, Policy_1.handleResultType)(MyError1)
            .orResultType(MyError2)
            .orResultType(MyError3, e => e.message === 'foo')
            .orWhenResult(e => e === 'potato'), { maxAttempts: 10 }).execute(fn)).to.equal('ok!');
        (0, chai_1.expect)(fn).to.have.callCount(5);
    });
    it('applies use', async () => {
        class Calculator {
            double(n, context) {
                if (context.attempt < 2) {
                    throw new Error('failed');
                }
                return { n: n * 2, ...context };
            }
        }
        __decorate([
            (0, Policy_1.usePolicy)((0, Policy_1.retry)(Policy_1.handleAll, { maxAttempts: 5 }))
        ], Calculator.prototype, "double", null);
        const c = new Calculator();
        // @ts-ignore
        const r = await c.double(2);
        (0, chai_1.expect)(r).to.deep.equal({
            n: 4,
            signal: r.signal,
            attempt: 2,
        });
    });
    it('uses cancellation token in use', async () => {
        class Calculator {
            double(n, context) {
                (0, chai_1.expect)(n).to.equal(2);
                (0, chai_1.expect)(context.signal.aborted).to.be.false;
                cts.abort();
                (0, chai_1.expect)(context.signal.aborted).to.be.true;
                return n * 2;
            }
        }
        __decorate([
            (0, Policy_1.usePolicy)((0, Policy_1.retry)(Policy_1.handleAll, { maxAttempts: 5 }))
        ], Calculator.prototype, "double", null);
        const cts = new AbortController();
        const c = new Calculator();
        // @ts-ignore
        (0, chai_1.expect)(await c.double(2, cts.signal)).to.equal(4);
    });
});
